Okta SDK Panics on bad URI or Token

### Expected Behavior

Nice pretty error when consumer provides an invalid Okta URI or Token. We should recover from this panic or even better, the panic never happen!

### Actual Behavior

PANIC at the disco. 

https://github.com/terraform-providers/terraform-provider-okta/blob/0c1611a960c7c7b35c14ca2edbf99649c970119a/vendor/github.com/okta/okta-sdk-golang/okta/okta.go#L87

### Steps to Reproduce

1. configure Okta provider with invalid params
2. run it against anything

ok, getting somewhere with this. my variable for the api was misconfigured. there is an error that is passed only sometimes in debug mode saying no api key found (how i noticed). is there a way to pass this to an actual TF error? when i hard coded the api in, it looks to be working, i'm now just getting a misformed body error.  here is the whole okta TF.
error:
 `2020/02/20 15:15:28 [ERROR] <root>: eval: *terraform.EvalSequence, err: The API returned an error: The request body was not well-formed.`

TF:
```
provider "okta" {
  org_name  = "dev-438690"
  base_url  = "okta.com"
  api_token = "xxxxxxxxx"
}

resource "okta_app_saml" "app-saml-dev" {
  label                    = "dev app"
  sso_url                  = "https://app-dev.comp.com/wg/saml/SSO/index.html"
  recipient                = "https://app-dev.comp.com/wg/saml/SSO/index.html"
  destination              = "https://app-dev.comp.com/wg/saml/SSO/index.html"
  audience                 = "https://app-dev.comp.com"
  subject_name_id_template = "$${user.userName}"
  subject_name_id_format   = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
  response_signed          = true
  assertion_signed         = true
  signature_algorithm      = "RSA_SHA256"
  digest_algorithm         = "SHA256"
  honor_force_authn        = true
  authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"

  attribute_statements {
    name         = "username"
    filter_value = "user.login+\".local\""
  }
}
```

--- a/examples/okta_app_saml/datasource.tf
+++ b/examples/okta_app_saml/datasource.tf
@@ -11,6 +11,13 @@ resource "okta_app_saml" "test" {
   digest_algorithm         = "SHA256"
   honor_force_authn        = false
   authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"
+
+  attribute_statements {
+    type         = "GROUP"
+    name         = "Attr Two"
+    filter_type  = "STARTS_WITH"
+    filter_value = "test"
+  }
 }
 
 data "okta_app_saml" "test" {

--- a/okta/app.go
+++ b/okta/app.go
@@ -145,47 +145,6 @@ var baseAppSwaSchema = map[string]*schema.Schema{
 	},
 }
 
-var attributeStatements = map[string]*schema.Schema{
-	"filter_type": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Description: "Type of group attribute filter",
-	},
-	"filter_value": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Description: "Filter value to use",
-	},
-	"name": {
-		Type:     schema.TypeString,
-		Required: true,
-	},
-	"namespace": {
-		Type:     schema.TypeString,
-		Optional: true,
-		Default:  "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified",
-		ValidateDiagFunc: stringInSlice([]string{
-			"urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified",
-			"urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
-			"urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
-		}),
-	},
-	"type": {
-		Type:     schema.TypeString,
-		Optional: true,
-		Default:  "EXPRESSION",
-		ValidateDiagFunc: stringInSlice([]string{
-			"EXPRESSION",
-			"GROUP",
-		}),
-	},
-	"values": {
-		Type:     schema.TypeList,
-		Optional: true,
-		Elem:     &schema.Schema{Type: schema.TypeString},
-	},
-}
-
 var appSamlDiffSuppressFunc = func(k, old, new string, d *schema.ResourceData) bool {
 	// Conditional default
 	return new == "" && old == "http://www.okta.com/${org.externalKey}"

--- a/okta/data_source_okta_app_saml.go
+++ b/okta/data_source_okta_app_saml.go
@@ -223,7 +223,38 @@ func dataSourceAppSaml() *schema.Resource {
 				Type:     schema.TypeList,
 				Optional: true,
 				Elem: &schema.Resource{
-					Schema: attributeStatements,
+					Schema: map[string]*schema.Schema{
+						"filter_type": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Type of group attribute filter",
+						},
+						"filter_value": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Filter value to use",
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The reference name of the attribute statement",
+						},
+						"namespace": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The name format of the attribute",
+						},
+						"type": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The type of attribute statements object",
+						},
+						"values": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
 				},
 			},
 			"single_logout_issuer": {

--- a/okta/data_source_okta_app_saml_test.go
+++ b/okta/data_source_okta_app_saml_test.go
@@ -52,5 +52,12 @@ resource "okta_app_saml" "test" {
   digest_algorithm         = "SHA256"
   honor_force_authn        = false
   authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"
+
+  attribute_statements {
+    type         = "GROUP"
+    name         = "Attr Two"
+    filter_type  = "STARTS_WITH"
+    filter_value = "test"
+  }
 }`, d)
 }

--- a/okta/resource_okta_app_saml.go
+++ b/okta/resource_okta_app_saml.go
@@ -39,6 +39,25 @@ func resourceAppSaml() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
+		CustomizeDiff: func(_ context.Context, d *schema.ResourceDiff, v interface{}) error {
+			jwks, ok := d.GetOk("attribute_statements")
+			if !ok {
+				return nil
+			}
+			for i := range jwks.([]interface{}) {
+				objType := d.Get(fmt.Sprintf("attribute_statements.%d.type", i)).(string)
+				if (d.Get(fmt.Sprintf("attribute_statements.%d.filter_type", i)).(string) != "" ||
+					d.Get(fmt.Sprintf("attribute_statements.%d.filter_value", i)).(string) != "") &&
+					objType != "GROUP" {
+					return errors.New("invalid 'attribute_statements': when setting 'filter_value' or 'filter_type', value of 'type' should be set to 'GROUP'")
+				}
+				if objType == "GROUP" &&
+					len(convertInterfaceToStringArrNullable(d.Get(fmt.Sprintf("attribute_statements.%d.values", i)))) > 0 {
+					return errors.New("invalid 'attribute_statements': when setting 'values', 'type' should be set to 'EXPRESSION'")
+				}
+			}
+			return nil
+		},
 		// For those familiar with Terraform schemas be sure to check the base application schema and/or
 		// the examples in the documentation
 		Schema: buildAppSchema(map[string]*schema.Schema{
@@ -276,7 +295,47 @@ func resourceAppSaml() *schema.Resource {
 				Type:     schema.TypeList,
 				Optional: true,
 				Elem: &schema.Resource{
-					Schema: attributeStatements,
+					Schema: map[string]*schema.Schema{
+						"filter_type": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							Description:      "Type of group attribute filter",
+							ValidateDiagFunc: stringInSlice([]string{"STARTS_WITH", "EQUALS", "CONTAINS", "REGEX"}),
+						},
+						"filter_value": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "Filter value to use",
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The reference name of the attribute statement",
+						},
+						"namespace": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified",
+							ValidateDiagFunc: stringInSlice([]string{
+								"urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified",
+								"urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
+								"urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
+							}),
+							Description: "The name format of the attribute",
+						},
+						"type": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							Default:          "EXPRESSION",
+							ValidateDiagFunc: stringInSlice([]string{"GROUP", "EXPRESSION"}),
+							Description:      "The type of attribute statements object",
+						},
+						"values": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
 				},
 			},
 			"single_logout_issuer": {

--- a/website/docs/d/app_saml.html.markdown
+++ b/website/docs/d/app_saml.html.markdown
@@ -96,13 +96,13 @@ data "okta_app_saml" "example" {
 
 - `acs_endpoints` - An array of ACS endpoints. You can configure a maximum of 100 endpoints.
 
-- `attribute_statements` - (Optional) List of SAML Attribute statements.
-  - `name` - (Required) The name of the attribute statement.
-  - `filter_type` - (Optional) Type of group attribute filter.
-  - `filter_value` - (Optional) Filter value to use.
-  - `namespace` - (Optional) The attribute namespace. It can be set to `"urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified"`, `"urn:oasis:names:tc:SAML:2.0:attrname-format:uri"`, or `"urn:oasis:names:tc:SAML:2.0:attrname-format:basic"`.
-  - `type` - (Optional) The type of attribute statement value. Can be `"EXPRESSION"` or `"GROUP"`.
-  - `values` - (Optional) Array of values to use.
+- `attribute_statements` - List of SAML Attribute statements.
+  - `name` - The name of the attribute statement.
+  - `filter_type` - Type of group attribute filter.
+  - `filter_value` - Filter value to use.
+  - `namespace` - The attribute namespace.
+  - `type` - The type of attribute statement value.
+  - `values` - Array of values to use.
 
 - `single_logout_issuer` - The issuer of the Service Provider that generates the Single Logout request.
 

--- a/website/docs/r/app_saml.html.markdown
+++ b/website/docs/r/app_saml.html.markdown
@@ -113,7 +113,7 @@ The following arguments are supported:
   - `filter_type` - (Optional) Type of group attribute filter. Valid values are: `"STARTS_WITH"`, `"EQUALS"`, `"CONTAINS"`, or `"REGEX"`
   - `filter_value` - (Optional) Filter value to use.
   - `namespace` - (Optional) The attribute namespace. It can be set to `"urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified"`, `"urn:oasis:names:tc:SAML:2.0:attrname-format:uri"`, or `"urn:oasis:names:tc:SAML:2.0:attrname-format:basic"`.
-  - `type` - (Optional) The type of attribute statement value. Can only be `"GROUP"`.
+  - `type` - (Optional) The type of attribute statement value. Valid values are: `"EXPRESSION"` or `"GROUP"`. Default is `"EXPRESSION"`.
   - `values` - (Optional) Array of values to use.
 
 - `key_years_valid` - (Optional) Number of years the certificate is valid (2 - 10 years).

--- a/website/docs/r/app_saml.html.markdown
+++ b/website/docs/r/app_saml.html.markdown
@@ -110,10 +110,10 @@ The following arguments are supported:
 
 - `attribute_statements` - (Optional) List of SAML Attribute statements.
   - `name` - (Required) The name of the attribute statement.
-  - `filter_type` - (Optional) Type of group attribute filter.
+  - `filter_type` - (Optional) Type of group attribute filter. Valid values are: `"STARTS_WITH"`, `"EQUALS"`, `"CONTAINS"`, or `"REGEX"`
   - `filter_value` - (Optional) Filter value to use.
   - `namespace` - (Optional) The attribute namespace. It can be set to `"urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified"`, `"urn:oasis:names:tc:SAML:2.0:attrname-format:uri"`, or `"urn:oasis:names:tc:SAML:2.0:attrname-format:basic"`.
-  - `type` - (Optional) The type of attribute statement value. Can be `"EXPRESSION"` or `"GROUP"`.
+  - `type` - (Optional) The type of attribute statement value. Can only be `"GROUP"`.
   - `values` - (Optional) Array of values to use.
 
 - `key_years_valid` - (Optional) Number of years the certificate is valid (2 - 10 years).

